File: src/main/java/com/uml/generator/generator/DiagramGenerator.java
package com.uml.generator.generator;

import com.uml.generator.model.Trace;
import java.util.List;

/**
 * Interface for UML diagram generators.
 */
public interface DiagramGenerator {

    /**
     * Generates XMI (XML Metadata Interchange) content from traces.
     * Conforms to OMG XMI 2.5.1 and UML 2.5.1 specifications.
     * 
     * @param traces list of traces to analyze
     * @return XMI XML content as string
     */
    String generateXmi(List<Trace> traces);

    /**
     * Gets the diagram type name.
     * 
     * @return diagram type (e.g., "sequence", "component", "deployment")
     */
    String getDiagramType();
}

File: python_version/jaeger_uml_generator/generators/diagram_generator.py
"""Abstract base class for diagram generators."""

from abc import ABC, abstractmethod
from typing import List
from ..models import Trace


class DiagramGenerator(ABC):
    """Abstract base class for all diagram generators."""
    
    @abstractmethod
    def get_diagram_type(self) -> str:
        """
        Get the type of diagram this generator creates.
        
        Returns:
            Diagram type name (e.g., 'sequence', 'component', 'deployment')
        """
        pass
    
    @abstractmethod
    def generate_xmi(self, traces: List[Trace]) -> str:
        """
        Generate XMI content for the given traces.
        
        Args:
            traces: List of Trace objects
            
        Returns:
            XMI content as a string
        """
        pass
